File: src/stages/writer.ts
import { writeFileSync, mkdirSync } from 'node:fs';
import { resolve, dirname, basename } from 'node:path';
import { log } from '../utils/logger.js';

export interface WriteOptions {
  sourcePath: string;
  outputPath?: string;
}

/**
 * Write Haskell source to a .hs file.
 */
export function writeHaskell(
  haskellSource: string,
  opts: WriteOptions
): string {
  const outPath = opts.outputPath
    ?? resolve(dirname(opts.sourcePath), basename(opts.sourcePath, '.soel') + '.hs');

  mkdirSync(dirname(outPath), { recursive: true });
  writeFileSync(outPath, haskellSource + '\n', 'utf-8');

  log.success(`Wrote ${outPath}`);
  return outPath;
}

/**
 * Write IR JSON to a file.
 */
export function writeIR(ir: unknown, sourcePath: string): string {
  const outPath = resolve(
    dirname(sourcePath),
    basename(sourcePath, '.soel') + '.ir.json'
  );

  mkdirSync(dirname(outPath), { recursive: true });
  writeFileSync(outPath, JSON.stringify(ir, null, 2) + '\n', 'utf-8');

  log.success(`Wrote ${outPath}`);
  return outPath;
}

File: hs/src/Soel/Stages/Writer.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}

module Soel.Stages.Writer
  ( WriteOptions(..)
  , writeHaskell
  , writeIR
  ) where

import Control.Monad.IO.Class (liftIO)
import Data.Aeson (Value, encode)
import qualified Data.ByteString.Lazy as BL
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import System.Directory (createDirectoryIfMissing)
import System.FilePath ((</>), takeDirectory, takeBaseName, replaceExtension)

import Soel.Types (App)
import Soel.Utils.Logger (logSuccess)

-- | Options for writing Haskell output.
data WriteOptions = WriteOptions
  { woSourcePath :: !FilePath
  , woOutputPath :: !(Maybe FilePath)
  } deriving (Show)

-- | Write Haskell source to a .hs file. Returns the output path.
writeHaskell :: Text -> WriteOptions -> App FilePath
writeHaskell haskellSource WriteOptions{..} = liftIO $ do
  let outPath = case woOutputPath of
        Just p  -> p
        Nothing -> replaceExtension woSourcePath ".hs"

  createDirectoryIfMissing True (takeDirectory outPath)
  TIO.writeFile outPath (haskellSource <> "\n")

  logSuccess $ "Wrote " <> T.pack outPath
  pure outPath

-- | Write IR JSON to a file. Returns the output path.
writeIR :: Value -> FilePath -> App FilePath
writeIR ir sourcePath = liftIO $ do
  let outPath = replaceExtension sourcePath ".ir.json"

  createDirectoryIfMissing True (takeDirectory outPath)
  BL.writeFile outPath (encode ir)

  logSuccess $ "Wrote " <> T.pack outPath
  pure outPath
